### PR TITLE
enhance(datastore): make scan atomic

### PR DESCRIPTION
**Pull request by xuchuan**
## Modules
- [ ] UI
- [X] Controller
- [ ] Agent
- [ ] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [ ] run code format and lint check
- [X] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java [hunk 1]
```diff
@@ -51,69 +54,178 @@ public void update(String tableName,
...
+                Map<String, ColumnType> columnTypeMap;
+            }
+
+            var tables = req.getTables().stream().map(info -> {
+                var ret = new TableMeta();
+                ret.tableName = info.getTableName();
+                ret.table = this.getTable(info.getTableName());
+                ret.schema = ret.table.getSchema();
+                ret.columns = this.getColumnAliases(ret.schema, info.getColumns());
+                ret.columnTypeMap = ret.schema.getColumnTypeMapping(ret.columns);
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java:131]: It seems that `keepNone` of tableInfo is useless?

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java [hunk 2]
```diff
@@ -51,69 +54,178 @@ public void update(String tableName,
...
+        var limit = req.getLimit();
+        if (limit > 1000) {
+            throw new SWValidationException(SWValidationException.ValidSubject.DATASTORE,
+                    "limit must be less or equal to 1000. request=" + req);
         }
-        var columnTypeMap = new HashMap<String, ColumnType>();
-        for (var it : iters) {
-            columnTypeMap.putAll(it.getColumnTypeMapping());
+        if (limit < 0) {
+            limit = 1000;
```
[the remark this answers is not in the document]
Comment (xuchuan) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java:102]: to get the table schema. added a test case for this scenario.

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java [hunk 3]
```diff
@@ -331,127 +335,77 @@ synchronized public RecordList query(
...
+            startKey = MemoryTableImpl.this.recordMap.ceilingKey(startKey);
+        } else {
+            startKey = MemoryTableImpl.this.recordMap.higherKey(startKey);
+        }
+        if (endKey == null) {
+            endKey = MemoryTableImpl.this.recordMap.lastKey();
+        } else if (endInclusive) {
+            endKey = MemoryTableImpl.this.recordMap.floorKey(endKey);
+        } else {
+            endKey = MemoryTableImpl.this.recordMap.lowerKey(endKey);
```
[the remark this answers is not in the document]
Comment (xuchuan) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java]: These codes come from the old implementation with the iterator. Now fixed.

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java (hunk 1 above)
Comment (xuchuan) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java:132]: Done

**Review by jialeicui — APPROVED** [review]
LGTM

**Pull request merged**